ee/agent/knapsack/knapsack.go

```diff
@@ -248,13 +248,88 @@ func (k *knapsack) CurrentEnrollmentStatus() (types.EnrollmentStatus, error) {
 	return types.Enrolled, nil
 }
 
+// SetEnrollmentDetails updates the enrollment details, merging with existing details
 func (k *knapsack) SetEnrollmentDetails(newDetails types.EnrollmentDetails) {
+	// Get current details or empty struct if nil
+	// capture old values for logging
+	var oldDetails types.EnrollmentDetails
+	if enrollmentDetails != nil {
+		oldDetails = *enrollmentDetails
+	}
+
+	// Merge the new details with existing ones
+	mergedDetails := mergeEnrollmentDetails(oldDetails, newDetails)
+
+	// Log old and merged details - simpler approach
 	k.Slogger().Log(context.Background(), slog.LevelDebug,
 		"updating enrollment details",
-		"old_details", fmt.Sprintf("%+v", enrollmentDetails),
-		"new_details", fmt.Sprintf("%+v", newDetails),
+		"old_details", fmt.Sprintf("%+v", oldDetails),
+		"new_details", fmt.Sprintf("%+v", mergedDetails),
 	)
-	enrollmentDetails = &newDetails
+
+	// Update the global enrollment details
+	enrollmentDetails = &mergedDetails
+}
+
+// mergeEnrollmentDetails combines old and new details, only updating non-empty fields
+func mergeEnrollmentDetails(oldDetails, newDetails types.EnrollmentDetails) types.EnrollmentDetails {
```

Comment:
Could we add a new `ee/agent/knapsack/knapsack_test.go` with a `Test_mergeEnrollmentDetails` unit test to test this function and prevent us from introducing regressions in the future?